rewrite initial sync

Glease opened the pull request:
the current implementation suffers a number of problems
1. spam packets on full sync (1 per slot) and this is not very good. it's not very significant tho
2. if a new player joins it will not receive baubles data of existing players. in effect, this is not a major issue (as of now) but the rendering of those would be missing
3. if a new player joins it will clear the local baubles data of everyone else on the server. in effect, your own baubles won't function (unless it works purely on server side and until you refetch those data by opening baubles inventory once) and will not be able to see anyone's baubles except the one just joined the server

this PR aims to address all of these

Regarding compatibility: I've inferred that network protocol is not something that should not be left unchanged, so I took the liberty of rewriting the initial sync completely. aside from protocol change, this should not affect compatibility with old baubles as long as both client and server is using BaublesExpanded of the same version

Fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/21682
Fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/21957
Fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/24370

Glease committed:
rewrite initial sync

DarkShadow44 commented on src/main/java/baubles/common/lib/PlayerHandler.java:
```diff
@@ -26,6 +26,10 @@ public static void clearClientPlayerBaubles() {
+    public static void clearClientPlayerBaubles(EntityPlayer player) {
+        playerBaublesClient.remove(player.getCommandSenderName());
+    }
+
```
Isn't this superfluous if we clear the entire map already?

Glease replied:
It's exactly clearing the entire map being problematic. full sync is being sent to everyone so whenever someone logins everyone else's map got cleared and left with one new entry of that person's baubles

DarkShadow44 replied:
Yeah, I understand the issue now. But isn't either clearing the whole map on server login or clearing the map for every packet enough? I mean, it doesn't hurt though.

DarkShadow44 commented on src/main/java/baubles/common/network/PacketSyncBauble.java:
```diff
@@ -60,20 +52,7 @@ public IMessage onMessage(PacketSyncBauble message, MessageContext ctx) {
-			else {
-				baubles.setInventorySlotContents(message.slot, message.bauble);
-			}
+            PlayerHandler.getPlayerBaubles(player).stackList[message.slot] = message.bauble;
```
Need to use `setInventorySlotContents` here, otherwise it won't sync to the client. You can test that by equipping a botania reach ring with right-click.

DarkShadow44 commented on src/main/java/baubles/common/event/EventHandlerNetwork.java:
```diff
@@ -23,9 +23,13 @@ public void playerLoggedInEvent(PlayerEvent.PlayerLoggedInEvent event)    {
 		if (side == Side.SERVER)        {
 			// Apply all baubles
             InventoryBaubles baubles = PlayerHandler.getPlayerBaubles(event.player);
+            PacketHandler.INSTANCE.sendToAll(new PacketSyncAllBauble(event.player));
```
This line is superfluous, since the current player is already part of the `playerEntityList` next line. It sends the packet twice, leading to baubles being applies twice (like reach ring)

Glease replied:
sigh.. I was meant to exempt the current player from full sync but I forgot. should add it back

DarkShadow44 commented:
>     1. spam packets on full sync (1 per slot) and this is not very good. it's not very significant tho

I don't think that's any problem, it's minor overhead. Not sure if it's worth to rewrite that and risk regressions. This would probably need some good testing, I already found two issues.

>     2. if a new player joins it will not receive baubles data of existing players. in effect, this is not a major issue (as of now) but the rendering of those would be missing

True. Are we sure that loading baubles from other players doesn't affect the current player? AFAIK they never were synced properly, so I'm a bit worried.

>     3. if a new player joins it will clear the local baubles data of everyone else on the server. in effect, your own baubles won't function (unless it works purely on server side and until you refetch those data by opening baubles inventory once) and will not be able to see anyone's baubles except the one just joined the server

Not sure what you mean, where does it clear the local bauble data of everyone else? Can you tell me how to reproduce the issue?

Glease commented:
> > ```
> > 1. spam packets on full sync (1 per slot) and this is not very good. it's not very significant tho
> > ```
> 
> I don't think that's any problem, it's minor overhead. Not sure if it's worth to rewrite that and risk regressions. This would probably need some good testing, I already found two issues.

this alone is not worth fixing. the more pressing issue the bottom one

> 
> > ```
> > 2. if a new player joins it will not receive baubles data of existing players. in effect, this is not a major issue (as of now) but the rendering of those would be missing
> > ```
> 
> True. Are we sure that loading baubles from other players doesn't affect the current player? AFAIK they never were synced properly, so I'm a bit worried.
> 
> > ```
> > 3. if a new player joins it will clear the local baubles data of everyone else on the server. in effect, your own baubles won't function (unless it works purely on server side and until you refetch those data by opening baubles inventory once) and will not be able to see anyone's baubles except the one just joined the server
> > ```
> 
> Not sure what you mean, where does it clear the local bauble data of everyone else? Can you tell me how to reproduce the issue?

1. start a server with requireGlasses on
2. have a player join. wear glasses
3. have another player join. boom. you can no longer see into inventories

![syncbug](https://github.com/user-attachments/assets/24da84e3-2237-4944-849c-0c749678a893)

I'm pretty surprised nobody reported this yet... and when asking delta folks they say it's working perfectly. I can only blame thermos for the moment

Glease committed:
do not send to same player twice

Glease committed:
revert revert to keep sending on equip event on client side

DarkShadow44 commented:
> > Not sure what you mean, where does it clear the local bauble data of everyone else? Can you tell me how to reproduce the issue?
> 
>     1. start a server with requireGlasses on
> 
>     2. have a player join. wear glasses
> 
>     3. have another player join. boom. you can no longer see into inventories

Oh right, I sent the initialization packet to all players instead of just the current one. I originally intended to only send it to the player currently logging in. Good catch!

DarkShadow44 commented on src/main/java/baubles/common/network/PacketSyncAllBauble.java:
```diff
@@ -0,0 +1,78 @@
+	public IMessage onMessage(PacketSyncAllBauble message, MessageContext ctx) {
+		World world = Baubles.proxy.getClientWorld();
+		if (world == null) return null;
+		Entity e = world.getEntityByID(message.playerId);
```
This fails in MP (all tested in dev with gradlew runServer and runClient):
Have two players online, one wearing a botania fluegel tiara (for visible wings).
Have the other player logout and back in, `Entity e` will be null here, the other player is not part of the world yet.
This will probably also break if the other user is in a different dimension.

The initial code never accounted for this case since sync on login is new, so we need to fix that as well. Currently not sure what the best  course of action to fix this would be.

Some initial thoughts:
Could put the message into a queue and apply it when the player exists, but that seems a big hacky. Especially since the player might be in a different dimension and log out there, leaving us with a message that can never be handled. Might need to handle dimension changed event on server for that to work properly.

Glease replied:
~~let's just send the player name instead~~ it doesn't really solve the problem that there is no player entity

Glease commented:
thanks for catching the case where player entity is not present. it's getting late where I live so I'll take a look tomorrow

Glease commented:
So I have been thinking about this for a while now. for now this is what I have in mind

1. the initial sync will no longer be a server wide broadcast. it'd be unicast to the one just logged in
2. sends initial sync from server side when it's being tracked by a player. Forge does have events for this: StartTracking and StopTracking in PlayerEvent
3. filters any other sync code to only send to player tracking the player who just had updated his baubles

PLASMAchicken commented:
I think this fixes this Issue right? https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/21957

DarkShadow44 commented:
> I think this fixes this Issue right? [GTNewHorizons/GT-New-Horizons-Modpack#21957](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/21957)

Yes, it should. But this PR needs a bit of a rework still.

Ethryan commented:
> thanks for catching the case where player entity is not present. it's getting late where I live so I'll take a look tomorrow

This still hasn’t been fixed, right?

DarkShadow44 commented:
> > thanks for catching the case where player entity is not present. it's getting late where I live so I'll take a look tomorrow
> 
> This still hasn’t been fixed, right?

Not fixed yet, no.

Dream-Master committed:
Merge branch 'gtnh' into fix/sync

Dream-Master committed:
Merge branch 'gtnh' into fix/sync

Dream-Master committed:
Merge branch 'gtnh' into fix/sync

boubou19 commented:
@Glease @DarkShadow44 any news here?

Luca-Guettinger commented on src/main/java/baubles/common/network/PacketSyncAllBauble.java:
```diff
@@ -0,0 +1,78 @@
+            PlayerHandler.clearClientPlayerBaubles(player);
+            InventoryBaubles baubles = PlayerHandler.getPlayerBaubles(player);
+            message.inventory.forEachEntry((a, b) -> {
+                baubles.stackList[a] = b;
```
Slot index `a` from the network is used directly as array index into stackList without bounds checking. If the server has more bauble slots than the client, or the packet is malformed, this crashes with ArrayIndexOutOfBoundsException. Should validate a >= 0 && a < baubles.getSizeInventory() before the write.

Glease replied:
slotLimit is a compile time constant. I'm not sure how this could happen

DarkShadow44 replied:
@Luca-Guettinger Do you have a way to reproduce such a crash?

Luca-Guettinger replied:
no its more of a defensive thing, if you think it cant happen resolve it. If a mod uses the api wrong and only registers bauble slots on client side.

Glease replied:
For future reference: the array in question is created by baubles only and its size is the fixed compile time constant `slotLimit`. While in theory stackList can be overwritten by a careless mod, we have more things to worry than streaming inventory content to client fails

Luca-Guettinger commented on src/main/java/baubles/common/network/PacketSyncAllBauble.java:
```diff
@@ -0,0 +1,78 @@
+		if (e instanceof EntityPlayer player) {
+            PlayerHandler.clearClientPlayerBaubles(player);
+            InventoryBaubles baubles = PlayerHandler.getPlayerBaubles(player);
+            message.inventory.forEachEntry((a, b) -> {
```
If onPlayerLoad throws for any bauble, the exception propagates out of forEachEntry and the rest of the inventory is silently skipped. Wrapping the lambda body in try-catch would prevent one broken item from killing the whole sync.

Luca-Guettinger reviewed (commented):
Noticed a couple of things on top of the existing discussion.

Glease committed:
fix almost everything
1. proper sync based on vanilla entity tracker
2. try catch around onPlayerLoad()
3. player reference updating and stale InventoryBaubles pruning

Dream-Master committed:
Merge branch 'master' into fix/sync

DarkShadow44 commented on src/main/java/baubles/common/lib/PlayerHandler.java:
```diff
@@ -12,22 +13,59 @@
+        baublesPrune(playerBaublesClient, true);
+    }
+
+    private static void baublesPrune(HashMap<String, InventoryBaubles> map, boolean isClient) {
```
Wouldn't it more elegant to handle this though events instead of every N ticks?

Glease replied:
the problem is there is no event for player entity removal, especially on client side. on server side you must use a combination of multiple events to catch all cases of player entity removal and I really don't want to go down that rabbit hole. on client side those events don't fire at all and the only other way is to sync the change via a network packet, which is obviously extremely wasteful

of course if you think mixin is the way...

DarkShadow44 replied:
We could lookup the `InventoryBaubles` using the string, then check `InventoryBaubles.player` and update it if it doesn't match the current player.

IMHO it's not a problem to have entries for a player in the map if the player is currently not online. For client side we could clear the map once a player logs out. For server we could also handle logout, if we really need.

Is there anything else we need to do? That should handle everything we need, right? This approach would avoid dealing with player entity removal as well.

DarkShadow44 reviewed (commented):
Looks pretty good, thanks! Only thing I dislike is the pruning logic, I wish we could handle this a different way. Do you think that's doable?

DarkShadow44 committed:
Merge remote-tracking branch 'origin/master' into sync

Dream-Master committed:
Merge branch 'master' into fix/sync
